### PR TITLE
Let's support naive Pipeline Parallelism

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Trying to load a model in a single device is cool, but what if we can split the model across multiple devices? 
Users will just have to pass a custom `device_map` when loading the model, and it should work out of the box.

This PR adds the support of "Sequential Parallel" - termed as naive Pipeline Parallelism as the real Pipeline parallelism involves dealing with multi-processing and gradients synchronisation that cannot be handled easily.

This PR depends on the following PRs:

- `accelerate`: https://github.com/huggingface/accelerate/pull/1172
- `peft`: https://github.com/huggingface/peft/pull/145

TODOs:

- users should NOT apply DP and/or DeepSpeed with this approach as it remains untested
- should we introduce multi-gpu tests that we optionally run?
- test with 8bit models
- update docs

cc @lvwerra @edbeeching

### CODE REVIEW

**Comment by younesbelkada** [comment]
Experiments of gpt-neo-1b int8 + peft multi-GPU : https://wandb.ai/distill-bloom/trl/runs/x3d6fig6?workspace=user-younesbelkada
Single GPU baseline with peft and int8: https://wandb.ai/distill-bloom/trl/runs/rgcqxtfd?workspace=user-younesbelkada

**Comment by younesbelkada** [comment]
Ran a DP script with `accelerate launch  gpt2-sentiment.py` to make sure nothing is broken in DP and seems to work like charm!
@lvwerra @edbeeching this is ready for review

File: examples/sentiment/scripts/gpt-neo-1b-multi-gpu/gpt-neo-1b_peft.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    return dict((key, [d[key] for d in data]) for key in data[0])
+
+
+# set seed before initializing value head for deterministic eval
+set_seed(config.seed)
+
+# Now let's build the model, the reference model, and the tokenizer.
+pretrained_model = AutoModelForCausalLM.from_pretrained(
+    config.model_name, load_in_8bit=True, device_map="balanced", max_memory={0: "800MB", 1: "800MB"}
+)
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply examples/sentiment/scripts/gpt-neo-1b-multi-gpu/gpt-neo-1b_peft.py:138]: Same with peft. We could just pass the configs as kwargs, right?

File: trl/models/modeling_value_head.py [hunk 2]
```diff
@@ -278,6 +308,31 @@ def post_init(self, state_dict):
...
+                new_output = ()
+                for output in outputs:
+                    if isinstance(output, torch.Tensor):
+                        new_output += (output.to(first_device),)
+                    else:
+                        new_output += (output,)
+                return new_output
+
+            self.register_forward_hook(set_device_hook)
+            self.is_sequential_parallel = True
```
Comment (lvwerra) [inline trl/models/modeling_value_head.py:334]: an explanation of what this what would be useful. maybe some comments :)

File: trl/trainer/ppo_config.py [hunk 3]
```diff
@@ -99,6 +101,7 @@ def __init__(
         accelerator_kwargs: Optional[dict] = {},
         tracker_project_name: Optional[str] = "trl",
         max_grad_norm: Optional[float] = None,
+        optimize_cuda_cache: Optional[bool] = False,
```
Comment (lvwerra) [inline trl/trainer/ppo_config.py:104]: are there drawbacks to setting it to true?

Comment (lvwerra) [reply trl/trainer/ppo_config.py:104]: also the order in the docstring and the kwargs is different, i think it's better to be consistent :)

**Review by lvwerra — COMMENTED** [review]
Looks good overall. One main things that I think we need to fix soon is the way different approaches are loaded (peft, PP, int8). This would also allow us to test compatibility of different methods at loading time. Loading a model twice is not very intuitive but we can fix this in a dedicated PR.

File: trl/trainer/ppo_config.py (hunk 3 above)
Comment (younesbelkada) [reply trl/trainer/ppo_config.py:104]: Fixed the order! 
The drawback is maybe about the computational time of the `step` function

File: trl/models/modeling_value_head.py [hunk 4]
```diff
@@ -223,6 +227,32 @@ def post_init(self, state_dict):
...
+                "cpu" in self.pretrained_model.hf_device_map.values()
+                or "disk" in self.pretrained_model.hf_device_map.values()
+            ):
+                raise ValueError(
+                    "The model is offloaded on CPU or disk - CPU & disk offloading is not supported for ValueHead models."
+                )
+
+            first_device = list(set(self.pretrained_model.hf_device_map.values()))[0]
+
+            self.v_head = self.v_head.to(first_device)
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/models/modeling_value_head.py:241]: Because the lm_head is usually on the first device, I modified a bit to use the lm_head device instead

File: trl/models/modeling_value_head.py [hunk 5]
```diff
@@ -223,6 +227,32 @@ def post_init(self, state_dict):
...
+        if hasattr(self.pretrained_model, "hf_device_map"):
+            if (
+                "cpu" in self.pretrained_model.hf_device_map.values()
+                or "disk" in self.pretrained_model.hf_device_map.values()
+            ):
+                raise ValueError(
+                    "The model is offloaded on CPU or disk - CPU & disk offloading is not supported for ValueHead models."
+                )
+
+            first_device = list(set(self.pretrained_model.hf_device_map.values()))[0]
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/models/modeling_value_head.py:239]: fixed in [b9f75eb](https://github.com/lvwerra/trl/pull/210/commits/b9f75ebe60ada0e6c514f72699cdda64300e7c4e)

File: trl/models/modeling_value_head.py (hunk 2 above)
Comment (younesbelkada) [reply trl/models/modeling_value_head.py]: Done!

File: examples/sentiment/scripts/gpt-neo-1b-multi-gpu/gpt-neo-1b_peft.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    task_type="CAUSAL_LM",
+)
+
+pretrained_model = prepare_model_for_int8_training(pretrained_model)
+pretrained_model = get_peft_model(pretrained_model, lora_config)
+
+model = AutoModelForCausalLMWithValueHead.from_pretrained(pretrained_model)
+
+model.gradient_checkpointing_disable = model.pretrained_model.gradient_checkpointing_disable
+model.gradient_checkpointing_enable = model.pretrained_model.gradient_checkpointing_enable
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply examples/sentiment/scripts/gpt-neo-1b-multi-gpu/gpt-neo-1b_peft.py:175]: Yes actually this is already done: https://github.com/lvwerra/trl/blob/main/trl/models/modeling_base.py#L64-L68

[the remark this answers is not in the document]
Comment (younesbelkada) [reply examples/sentiment/scripts/gpt-neo-1b-multi-gpu/gpt-neo-1b_peft.py:175]: Removed these lines

**Review by lvwerra — APPROVED** [review]
